plugins/utils/auto-save.nix

```diff
@@ -6,139 +6,173 @@
   ...
 }:
 with lib;
-let
-  cfg = config.plugins.auto-save;
-in
-{
-  options.plugins.auto-save = helpers.neovim-plugin.extraOptionsOptions // {
-    enable = mkEnableOption "auto-save";
-
-    package = helpers.mkPluginPackageOption "auto-save" pkgs.vimPlugins.auto-save-nvim;
-
-    keymaps = {
-      silent = mkOption {
-        type = types.bool;
-        description = "Whether auto-save keymaps should be silent.";
-        default = false;
-      };
-
-      toggle = helpers.mkNullOrOption types.str "Keymap for running auto-save.";
-    };
-
-    enableAutoSave = helpers.defaultNullOpts.mkBool true ''
-      Whether to start auto-save when the plugin is loaded.
+helpers.neovim-plugin.mkNeovimPlugin config {
+  name = "auto-save";
+  originalName = "auto-save.nvim";
+  defaultPackage = pkgs.vimPlugins.auto-save-nvim;
+
+  maintainers = [ helpers.maintainers.braindefender ];
+
+  # TODO: introduced 2024-06-21, remove after 24.11
+  deprecateExtraOptions = true;
+  optionsRenamedToSettings = [
+    [
+      "executionMessage"
+      "message"
+    ]
+    [
+      "executionMessage"
+      "dim"
+    ]
+    [
+      "executionMessage"
+      "cleaningInterval"
+    ]
+    "triggerEvents"
+    "writeAllBuffers"
+    "debounceDelay"
+  ];
+
+  imports =
+    let
+      basePluginPath = [
+        "plugins"
+        "auto-save"
+      ];
+      settingsPath = basePluginPath ++ [ "settings" ];
+    in
+    [
+      (mkRenamedOptionModule (basePluginPath ++ [ "enableAutoSave" ]) (settingsPath ++ [ "enabled" ]))
+      (mkRemovedOptionModule (basePluginPath ++ [ "keymaps" ]) ''
+        Use the top-level `keymaps` option to create a keymap that runs :ASToggle
+
+        keymaps = [
+          { key = "<leader>s"; action = "<cmd>ASToggle<CR>"; }
+        ];
+      '')
+    ];
+
+  settingsOptions = {
+    enabled = helpers.defaultNullOpts.mkBool true ''
+      Whether to start auto-save when th plugin is loaded.
     '';
 
-    executionMessage = {
+    execution_message = {
+      enabled = helpers.defaultNullOpts.mkBool true ''
+        Show execution message after successful auto-save.
+      '';
+
       message =
-        helpers.defaultNullOpts.mkStr
-          {
-            __raw = ''
-              function()
-                return ("AutoSave: saved at " .. vim.fn.strftime("%H:%M:%S"))
-              end
-            '';
-          }
+        helpers.defaultNullOpts.mkLuaFn
```

Comment:
I think this change is wrong, [in the source](https://github.com/okuuva/auto-save.nvim/blob/5fe9ab0c42f0457f2a973e814a6352b8eeb04730/lua/auto-save/config.lua#L8) it is annotated as `string|fun():string`, so we want `mkStr` with a `__raw` plugin-default (& test), as it was before.